api/API_VERSIONING.md

```diff
@@ -0,0 +1,203 @@
+annotated as draft or experimental are exempt from this policy. The following annotations may be
+applied to a protobuf file, message, enum, field or enum value:
+* `[#proto-status: experimental]` indicating that backwards incompatible changes are likely and
+  major restructuring is anticipated.
+* `[#proto-status: draft]` indicating that backwards incompatible changes are possible but the
+  entity is near finalized and some implementation exists.
+* `[#not-implemented-hide:]` indicating that the entity is not implemented in Envoy and the entity
+  should be hidden from the Envoy documentation. No backwards incompatible changes are allowed.
+* `[#not-implemented-warn:]` indicating that the entity is not implemented in Envoy and the entity
+  should be included in the Envoy documentation with a warning. No backwards incompatible changes
+  are allowed.
+
+Note that changes to default values for wrapped types, e.g. `google.protobuf.UInt32Value` are not
+governed by the above policy. Any management server requiring stability across Envoy API or
+implementations within a major version should set explicit values for these fields.
+
+# API lifecycle
+
+The API lifecycle follows a calendar clock. At the end of Q3 each year, a major API version
+increment may occur for any Envoy API package, in concert with the quarterly Envoy release.
+
+Envoy will support at most three major versions of any API package at all times:
+* The current stable major version, e.g. v3.
+* The previous stable major version, e.g. v2. This is needed to ensure that we provide at least 1
+  year for a supported major version to sunset. By supporting two stable major versions
+  simultaneously, this makes it easier to coordinate control plane and Envoy rollouts as well. This
+  previous stable major version will be supported for at most 1 year.
+* Optionally, the next experimental alpha major version, e.g. v4alpha. This is a release candidate
+  for the next stable major version. This is only generated when the current stable major version
+  requires a breaking change at the next cycle, e.g. a deprecation or field rename. This release
+  candidate is mechanically generated via the
+  [protoxform](https://github.com/envoyproxy/envoy/tree/master/tools/protoxform) tool from the
+  current stable major version, making use of annotations such as `deprecated = true`. This is not a
+  human editable artifact.
+
+An example of how this might play out is that at the end of September in 2020, we will freeze
+`envoy.config.bootstrap.v4alpha` and this package will become the current stable major version
+`envoy.config.bootstrap.v4`. The `envoy.config.bootstrap.v3` package will become the previous stable
+major version and support for `envoy.config.bootstrap.v2` will be dropped from the Envoy
+implementation. Note that some transitively referenced package, e.g.
+`envoy.config.filter.network.foo.v2` may remain at version 2 during this release, if no changes were
```

Comment:
Is there any reason not to bump the version number of all packages when we introduce a new major version, even if the protos themselves did not change?  It seems like that might make the dependency story a bit easier to deal with.

Reply:
Yeah. I think this will be too cumbersome to management server vendors. We have O(100) protos. If we ask management server vendors to rewrite their logic for O(100) protos on each major version upgrade, this is a lot more expensive than rewriting for O(10) protos.

Within Envoy, we plan on doing some fancy reflection gymnastics to do runtime rewriting from `v2` to `v3alpha` (I'm working on this right now), when API messages are mostly unchanged. Internally, Envoy will always consume from the latest API, e.g. `v3alpha`. I think it's not reasonable to expect every management server to invent clever techniques like this though, we should assume the bulk of them will cope with churn via manual toil.